internal/charmstore/debug.go

```diff
@@ -15,8 +18,75 @@ func serveDebugInfo(http.Header, *http.Request) (interface{}, error) {
 	return appver.VersionInfo, nil
 }
 
-func newServiceDebugHandler() http.Handler {
+// GET /debug/check.
+func debugCheck(checks map[string]func() error) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+		n := len(checks)
+		type result struct {
+			name string
+			err  error
+		}
+		c := make(chan result)
+		for name, check := range checks {
+			go func(n string, check func() error) {
+				c <- result{name: n, err: check()}
+			}(name, check)
+		}
+		results := make(map[string]string, n)
+		var failed bool
+		for n > 0 {
+			res := <-c
+			if res.err == nil {
+				results[res.name] = "OK"
+			} else {
+				failed = true
+				results[res.name] = res.err.Error()
+			}
+			n--
+		}
+		w.Header().Set("Content-Type", "application/json")
+		if failed {
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+		enc := json.NewEncoder(w)
+		if err := enc.Encode(results); err != nil {
```

Comment:
This is pretty academic as it's really only useful for things that are really only going to look at the status code anyway. But if we are producing a different output for the errors then I think we should produce no output on success. Humans should use /v1/debug/status.